src/Compatibility/Microsoft.DotNet.ApiCompat.Shared/ValidateAssemblies.cs

```diff
@@ -0,0 +1,144 @@
+using Microsoft.DotNet.ApiCompatibility.Logging;
+using Microsoft.DotNet.ApiCompatibility.Runner;
+
+namespace Microsoft.DotNet.ApiCompat
+{
+    [ServiceProvider]
+    [Singleton(typeof(ISuppressionEngine), Factory = nameof(SuppressionEngineFactory))]
+    [Singleton(typeof(ICompatibilityLogger), Factory = nameof(LogFactory))]
+    [Singleton(typeof(IApiComparerFactory), typeof(ApiComparerFactory))]
+    [Singleton(typeof(IAssemblySymbolLoaderFactory), typeof(AssemblySymbolLoaderFactory))]
+    [Singleton(typeof(IMetadataStreamProvider), typeof(MetadataStreamProvider))]
+    [Singleton(typeof(IApiCompatRunner), typeof(ApiCompatRunner))]
+    internal partial class ValidateAssembliesServiceProvider
+    {
+        public Func<ICompatibilityLogger> LogFactory { get; }
+
+        public Func<ISuppressionEngine> SuppressionEngineFactory { get; }
+
+        public ValidateAssembliesServiceProvider(Func<ISuppressionEngine, ICompatibilityLogger> logFactory,
+            Func<ISuppressionEngine> suppressionEngineFactory)
+        {
+            // It's important to use GetService<T> here instead of directly invoking the factory
+            // to avoid two suppression engine being created.
+            LogFactory = () => logFactory(GetService<ISuppressionEngine>());
+            SuppressionEngineFactory = suppressionEngineFactory;
+        }
+    }
+
+    internal static class ValidateAssemblies
+    {
+        public static void Run(Func<ISuppressionEngine, ICompatibilityLogger> logFactory,
+            bool generateSuppressionFile,
+            string? suppressionFile,
+            string? noWarn, 
+            string[] leftAssemblies,
+            string[] rightAssemblies,
+            bool enableStrictMode,
+            string[][]? leftAssembliesReferences,
+            string[][]? rightAssembliesReferences,
+            bool createWorkItemPerAssembly,
+            (string CaptureGroupPattern, string ReplacementString)[]? leftAssembliesTransformationPatterns,
+            (string CaptureGroupPattern, string ReplacementString)[]? rightAssembliesTransformationPatterns)
+        {
+            // Configure the suppression engine. Ignore the passed in suppression file if it should be generated and doesn't yet exist.
+            string? suppressionFileForEngine = generateSuppressionFile && !File.Exists(suppressionFile) ? null : suppressionFile;
+            Func<ISuppressionEngine> suppressionEngineFactory = () => new SuppressionEngine(suppressionFileForEngine, noWarn, generateSuppressionFile);
+
+            // Initialize the service provider
+            ValidateAssembliesServiceProvider serviceProvider = new(logFactory, suppressionEngineFactory);
+
+            IApiCompatRunner apiCompatRunner = serviceProvider.GetService<IApiCompatRunner>();
+            ApiCompatRunnerOptions apiCompatOptions = new(enableStrictMode);
+
+            // Optionally provide a string transformer if a transformation pattern is passed in.
```

Comment:
As suggested on the offline code review, I think we should have a fallback either here or in the targets to use a regex pattern that matches a vanilla project's output location.

Reply:
That makes sense for the future but at this point we don't want to expose the ValidateAssemblies targets publicly, they will only reside in the transport package. We should see if it's possible to add a default in the future when this logic will be exposed more publicly.

What makes this problematic is that the counterpart frontend (console / global tool) doesn't have any information of a project's output path and hence will always require a manual user input.

I filed https://github.com/dotnet/sdk/issues/26860 to track that.